plugins/API/API.php

```diff
@@ -129,6 +129,14 @@ public function getSegmentsMetadata($idSites = array(), $_hideImplementationData
                     $segment->setPermission($isAuthenticatedWithViewAccess);
                 }
 
+                if ($segment->getSqlSegment() && $segment->getUnionOfSegments()) {
+                    throw new \Exception(sprintf('Union of segments and SQL segment is set for segment "%s", use only one of them', $segment->getName()));
+                }
+
+                if ($segment->getUnionOfSegments() && in_array($segment->getSegment(), $segment->getUnionOfSegments(), true)) {
+                    throw new \Exception(sprintf('The segment %s contains a union segment to itself', $segment->getName()));
+                }
```

Comment:
The alternative would be to put it in `Piwik/Segment` in core. I can move it there as we also check for a permission there already. It will be harder though to get notified in case of a misconfiguration as one only gets notified when actually applying the segment.  

The problem is that Segments are implemented wrong. It's a huge mix of things in core and plugins. If core worked actually with the raw segment classes and a raw list of segments and not the plugin metadata it would be quite easy. We cannot easily change this for now though. I can add it to the class you suggested but doesn't really belong there and we should maybe refactor it at some point